driver-core/src/main/com/mongodb/internal/connection/AbstractMultiServerCluster.java

```diff
@@ -367,7 +365,7 @@ private void invalidateOldPrimaries(final ServerAddress newPrimary) {
                 if (LOGGER.isInfoEnabled()) {
                     LOGGER.info(format("Rediscovering type of existing primary %s", serverTuple.description.getAddress()));
                 }
-                serverTuple.server.invalidate(new MongoNotPrimaryException(new BsonDocument(), serverTuple.description.getAddress()));
+                serverTuple.server.invalidate();
```

Comment:
According to SDAM:

> A note on invalidating the old primary: when a new primary is discovered, the client finds the previous primary (there should be none or one) and replaces its description with a default ServerDescription of type "Unknown." A multi-threaded client MUST request an immediate check for that server as soon as possible.

> If the old primary server version is 4.0 or earlier, the client MUST clear its connection pool for the old primary, too: the connections are all bad because the old primary has closed its sockets. If the old primary server version is 4.2 or newer, the client MUST NOT clear its connection pool for the old primary.

And that's what `invalidate` now does.  It doesn't make sense to forge an exception here, so got rid of that.  That lets us simplify the `ClustererableServer` API

Reply:
Does invalidate request an immediate check like the spec says in your linked comment?

Reply:
Yes